### ISSUE TITLE
limit() is not working with delete()

### ISSUE THREAD

**Issue by kilobyte2007** [knex/knex#6074]
# Environment

Knex version: 3.1.0
Database + version: MySQL 8.3.0
OS: Linux/MacOS

It seems like knex still doesn't support limit when using `delete`.

`.where(...).limit(100).delete()`

I see there's an older issue #96 about this which is closed as fixed but it still doesn't work.
One of the reasons stated in the original issue is the consistency across platforms. But from what I see in the docs - there's lots of things that are platform specific so I think it would make sense to implement this.

**Comment by jokeyrhyme** [comment]
i just encountered this with PostgreSQL, where DELETE does not support LIMIT at all, but the `knex` query builder happily accepts `.delete(...).limit(...)` and simply removes the LIMIT from the query before execution

very surprising! :)

**Comment by mercmobily** [comment]
I will get to this today or tomorrow

**Comment by mercmobily** [comment]
As a favour, would you be able to do a search in the issue list (I know, it's a little big...) and see if there are duplicate issues saying the same thing?

**Comment by ZeRego** [comment]
@mercmobily no issue with the keyword `limit` seems to match this scenario.

**Comment by jokeyrhyme** [comment]
I have 2 ideas for how to improve this, but they are probably both breaking changes

- modify the TypeScript types so that it's a compiler error to try to use .limit() on a DELETE query builder (but this wouldn't be appropriate if other DBs do support this)

- throw an error if the DB is Postgres and there's an attempt to produce a `DELETE ... LIMIT ...` query

in my use case, I was attempting to delete a test batch of 8K rows, and knex silently discarded the LIMIT to produce a valid DELETE query, result in a query that deleted 45M rows instead (causing a production outage due to a combination of other queries running at the same time)

not blaming knex for the outage: it was my code and i should have understood and tested the output of the query builder and it's ramifications 

but it was indeed surprising :)

**Comment by mercmobily** [comment]
> in my use case, I was attempting to delete a test batch of 8K rows, and knex silently discarded the LIMIT to produce a valid DELETE query, result in a query that deleted 45M rows instead (causing a production outage due to a combination of other queries running at the same time)

Oh my God... I can't even imagine.
In any case, I believe dropping the limit is potentially deadly (you can confirm that...) and it will need to be addressed; I am thinking of throwing, frankly.  It was my thought before, and it's even stronger now!

**Comment by myndzi** [comment]
It seems like sqlite and MS SQL server supports delete with limit (top), and the rest of the supported databases don't. However, I don't see in the code that any of them _use_ it. I can likely add this on to https://github.com/knex/knex/pull/6314 until something better is in place.

As I was developing that PR, I was thinking that there's a need to "ensure that all the data that the user supplied for building a query is actually used", which would address this kind of problem in a more complete fashion, but would be a deeper change.

**Comment by jokeyrhyme** [comment]
yeah, it's tricky

ideally, you wouldn't want consumers of knex to have to care too much about the specific database/backend

and as a static types advocate, ideally, I'd want the type system to tell me I'm attempting to do e.g. a MySQL query that won't work at all on PostgreSQL 

really hard to tie all that together in a way that doesn't over complicate the library or ruin usability

**Comment by myndzi** [comment]
I agree, but knex's API doesn't provide for a good way to do that. We'd need to know the statement type up front / first to be able to exclude the clause types from it. It's possible to do it the other way around, and like, "collect a list of the clauses in use" and then filter the methods available according to that, but it would be both more confusing as a user and pretty complex to implement in separation from the actual code. I'd call a type-based solution fodder for a future native-typescript version of the library.

In the interim, I have a stepwise improvement that simply throws when a user does a thing that might lose them data (based on an explicit list of incompatible statement/clause combinations). And there's a chance for an internal-only refactor down the road doing what I described: the query builder builds up an object that "knows" everything the user asked for. The query compiler "consumes" data from that object. If the object is non-empty after compiling the query, we throw an error: in these cases, we know that the user asked for something that the implementation does not account for. 

That solution would ensure that we _can't_ have this sort of issue, but it's a more comprehensive refactor, and I don't think it's a good idea to throw that in the mix at this stage when we're trying to just triage critical bugs and I'm trying to keep PRs well-isolated

**Comment by erulabs** [comment]
A bit confused here -- mysql8 _does_ support DELETE WHERE... See https://dev.mysql.com/doc/refman/8.0/en/delete.html

The fix merged in https://github.com/knex/knex/pull/6314 now breaks a feature that did work properly in mysql, but was a silent bug in postgres.
